pkg/webhook/subscription.go

```diff
@@ -0,0 +1,64 @@
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	opv1a1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/red-hat-storage/ocs-client-operator/api/v1alpha1"
+	"github.com/red-hat-storage/ocs-client-operator/pkg/utils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type SubscriptionAdmission struct {
+	Client  client.Client
+	Decoder *admission.Decoder
+	Log     logr.Logger
+}
+
+func (s *SubscriptionAdmission) Handle(ctx context.Context, req admission.Request) admission.Response {
+	s.Log.Info("Request received for admission review")
+
+	// review should be for a subscription
+	subscription := &opv1a1.Subscription{}
+	if err := s.Decoder.Decode(req, subscription); err != nil {
+		s.Log.Error(err, "failed to decode admission review as subscription")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("only subscriptions admission reviews are supported: %v", err))
+	}
+
+	// review should be for ocs-client-operator subscription
+	if subscription.Spec.Package != "ocs-client-operator" {
+		s.Log.Info("subscription package is not 'ocs-client-operator'")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("only ocs-client-operator subscription validation is supported"))
+	}
+
+	storageClients := &v1alpha1.StorageClientList{}
+	if err := s.Client.List(ctx, storageClients); err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed to list storageclients for validating subscription request: %v", err))
+	}
+
+	// this is the channel that user want to subscribe to
+	requestedUserSubcriptionChannel := subscription.Spec.Channel
+	for idx := range storageClients.Items {
+		storageClient := &storageClients.Items[idx]
+		namespacedName := client.ObjectKeyFromObject(storageClient)
+
+		annotations := storageClient.GetAnnotations()
+		if annotations == nil {
+			s.Log.Info(fmt.Sprintf("Rejecting review as storageclient %q doesn't have annotations set", namespacedName))
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("storageclient %q doesn't have annotations to refer subscription channel", namespacedName))
+		} else if allowedSubscriptionChannel, exist := annotations[utils.DesiredSubscriptionChannelAnnotationKey]; !exist {
+			s.Log.Info(fmt.Sprintf("Rejecting review as storageclient %q doesn't have desired subscription channel annotation", namespacedName))
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("storageclient %q doesn't has required subscription channel annotation", namespacedName))
+		} else if allowedSubscriptionChannel != requestedUserSubcriptionChannel {
```

Comment:
I think both these cases are wrong. If annotations do not exist or the specific annotation does not exist it means we have no restrictions on the allowed channel. In that case, we are returning an `admission.Errored`

Reply:
done.